### DISCUSSION TITLE
Move plugin loader class to separate repository

### DISCUSSION THREAD

**Discussion by marip8** [tesseract-robotics/tesseract#778] [Ideas]
I extracted and simplified the plugin loader from `tesseract` into a [separate repository](https://github.com/marip8/plugin_loader) so I can use it for other projects. We should consider replacing the version in `tesseract` with this version. I think the remaining changes that would be required are to ensure build on Windows, add changes to support Python bindings, and move the repo off my fork into `ros-industrial` or `swri-robotics`

**Comment by Levi-Armstrong** [comment]
We will most likely have to change the name to release on the ros build farm. Maybe boost_plugin_loader?

**Reply by Levi-Armstrong** [reply]
I think this could also be hosted in the tesseract robotics organization. Either way I do not particularly care where it lives as long as I am one of the maintainers since this core to tesseract functionality. I think we should move fast on this and get it moved where it will live and released on the build farm.

**Reply by marip8** [reply]
That name change sounds good to me. I also don't really care where it lives either. Let's get it fully ready first, and then we can move the repository to its official home

**Reply by Levi-Armstrong** [reply]
Sound good. I will start creating PRs. I will create an issue with a list of things we want to do.

**Comment by Levi-Armstrong** [comment]
There appears to be some missing function that are in the Tesseract version that I cannot find in the plugin loader class.

``` c++ 
/**
   * @brief Check if plugin is available
   * @param plugin_name The plugin name to find
   * @return True if plugin is found
   */
  inline bool isPluginAvailable(const std::string& plugin_name) const;

  /**
   * @brief Get the available plugins under the provided section
   * @param section The section name to get all available plugins
   * @return A list of available plugins under the provided section
   */
  inline std::vector<std::string> getAvailablePlugins(const std::string& section) const;

  /**
   * @brief Get the available sections within the provided search libraries
   * @return A list of available sections
   */
  inline std::vector<std::string> getAvailableSections(bool include_hidden = false) const;

  /**
   * @brief The number of plugins stored. The size of plugins variable
   * @return The number of plugins.
   */
  inline int count() const;
```

**Reply by marip8** [reply]
These should be easy enough to add. I didn't have a personal use-case for these, so I didn't include them initially.

**Comment by Levi-Armstrong** [comment]
The plugin code has been moved to this [repository](https://github.com/tesseract-robotics/boost_plugin_loader) and released on the ros build farm. Also debian and nuget packages have been uploaded to the [release](https://github.com/tesseract-robotics/boost_plugin_loader/releases).
